crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,194 @@
+use crate::ast::helpers::find_names;
+use crate::ast::types::{Node, Range};
+use crate::ast::visitor;
+use crate::ast::visitor::Visitor;
+use crate::checkers::ast::Checker;
+use crate::registry::Diagnostic;
+use crate::source_code::Locator;
+use crate::violation::Violation;
+use ruff_macros::{define_violation, derive_message_formats};
+use rustpython_parser::ast::{Expr, Stmt, StmtKind, Withitem};
+use std::iter::zip;
+
+define_violation!(
+    /// ## What it does
+    /// Checks for variables defined in `for` loops and `with` statements that get overwritten
+    /// within the body, for example by another `for` loop or `with` statement or by direct
+    /// assignment.
+    ///
+    /// ## Why is this bad?
+    /// Redefinition of a loop variable inside the loop's body causes its value to differ from
+    /// the original loop iteration for the remainder of the block, in a way that will likely
+    /// cause bugs.
+    ///
+    /// In Python, unlike many other languages, `for` loops and `with` statements don't define
+    /// their own scopes. Therefore, a nested loop that uses the same target variable name as
+    /// an outer loop will reuse the same actual variable, and the value from the last
+    /// iteration will "leak out" into the remainder of the enclosing loop.
+    ///
+    /// While this mistake is easy to spot in small examples, it can be hidden in larger
+    /// blocks of code where the definition and redefinition of the variable may not be
+    /// visible at the same time.
+    ///
+    /// ## Example
+    /// ```python
+    /// for i in range(10):
+    ///     i = 9
+    ///     print(i)  # prints 9 every iteration
+    ///
+    /// for i in range(10):
+    ///     for i in range(10):  # original value overwritten
+    ///         pass
+    ///     print(i)  # also prints 9 every iteration
+    ///
+    /// with path1.open() as f:
+    ///     with path2.open() as f:
+    ///         f = path2.open()
+    ///     print(f.readline())  # prints a line from path2
+    /// ```
+    pub struct RedefinedLoopName {
+        pub name: String,
+    }
+);
+impl Violation for RedefinedLoopName {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedefinedLoopName { name } = self;
+        format!("For loop or with statement variable `{name}` overwritten in body")
```

Comment:
Could we make the "binding kind" a parameter on `RedefinedLoopName`, and use it to narrow this message? (`DeferralKeyword` in `crates/ruff/src/rules/pyflakes/rules/yield_outside_function.rs` would be an example of what that might look like.)

Reply:
Working on this now.

Reply:
Done. Example messages:

```
test.py:2:9: PLW2901 Outer for loop variable `i` overwritten by for loop target with same name
test.py:6:9: PLW2901 Outer with statement variable `i` overwritten by for loop target with same name
test.py:12:5: PLW2901 Outer for loop variable `i` overwritten by assignment target with same name
```